File: gmrpc_lua/rpc_manager.lua
-- @hzyuxin
-- RPC Manager for handle rpc dispatch
local msgpack = require 'src/third/MessagePack'

RPC_Manager = luaClass()

function RPC_Manager:ctor()
end

function RPC_Manager:create()
	return RPC_Manager()
end

function RPC_Manager:register(rpc_handler)
	-- 注册一个rpc到rpc manager
	self[rpc_handler.eid] = rpc_handler
end

-- 消息转发，把对应的消息，发到对应的rpc_handler，然后进行执行
function RPC_Manager:call(eid, fun_name, ...)
	self[eid][fun_name](self[eid], ...)
end

function RPC_Manager:on_data(msg) 
	local unpack_msg = msgpack.unpack(msg)
	self:call(unpack(unpack_msg))
end

File: gmrpc_py/rpc_manager.py
__author__ = 'hzyuxin'

import msgpack
class RPC_Manager(object):
    def __init__(self):
        pass

    def register(self, rpc_handler):
        setattr(self, rpc_handler.eid, rpc_handler)

    def call(self, eid, fun_name, *param):
        getattr(getattr(self, eid), fun_name)(*param)

    def on_data(self, data):
        unpack_msg = msgpack.unpackb(data)
        self.call(*unpack_msg)

if __name__ == "__main__":
    def bar(eid, fun_name, arg1, arg2, arg3):
        print eid, fun_name, arg1, arg2, arg3
    def foo(eid, fun_name, *param):
        bar(eid, fun_name, *param)
    foo(*[1,2,3,4,5])
